deltacat/compute/compactor/steps/repartition.py

```diff
@@ -0,0 +1,132 @@
+import ray
+from deltacat import logs
+from deltacat.compute.compactor import DeltaAnnotated
+from deltacat.compute.compactor.model.repartition_result import RePartitionResult
+from deltacat.storage import interface as unimplemented_deltacat_storage
+from deltacat.storage import Partition
+from deltacat.utils.ray_utils.runtime import (
+    get_current_ray_task_id,
+    get_current_ray_worker_id,
+)
+from deltacat.utils.common import ReadKwargsProvider
+from deltacat.utils.performance import timed_invocation
+from deltacat.utils.metrics import emit_timer_metrics, MetricsConfig
+from deltacat.storage import Delta
+
+if importlib.util.find_spec("memray"):
+    import memray
+
+logger = logs.configure_deltacat_logger(logging.getLogger(__name__))
+
+
+def _timed_repartition(
+    annotated_delta: DeltaAnnotated,
+    repartition_type: str,
+    repartition_args: dict,
+    destination_partition: Partition,
+    enable_profiler: bool,
+    read_kwargs_provider: Optional[ReadKwargsProvider],
+    deltacat_storage=unimplemented_deltacat_storage,
+) -> RePartitionResult:
+
+    if repartition_type == "range":
+        # A delta that is partitioned by range is partitioned in such a way
+        # that each partition contains rows for which the partitioning expression value lies within a given range for a specified column.
+        # For example, if the partitioning column is "date", the partition ranges are "2020-01-01" to "2020-01-10" and "2020-01-11" to "2020-01-20".
+        # The partitioned delta will have two partitions, one for each range.
+        column: str = repartition_args["column"]
+        partition_ranges: List = repartition_args["ranges"]
+        task_id = get_current_ray_task_id()
+        worker_id = get_current_ray_worker_id()
+        with memray.Tracker(
+            f"repartition_{worker_id}_{task_id}.bin"
+        ) if enable_profiler else nullcontext():
+            tables: List[pa.Table] = deltacat_storage.download_delta(
+                annotated_delta,
+                storage_type=StorageType.LOCAL,
+                file_reader_kwargs_provider=read_kwargs_provider,
+            )
+            partitioned_tables = {i: [] for i in range(len(partition_ranges) + 1)}
+            total_record_count = 0
+            col_name_int64 = f"{column}_int64"
+            for table in tables:
+                total_record_count += len(table)
+                table_new = table.add_column(
+                    0,
+                    pa.field(col_name_int64, pa.int64()),
+                    pc.cast(table[column], pa.int64()),
+                )
+                for filter_value in partition_ranges:
+                    partitioned_tables[partition_ranges.index(filter_value)].append(
+                        table_new.filter(
+                            (pc.field(col_name_int64) <= pc.scalar(filter_value))
+                        )
+                    )
+                    if filter_value == partition_ranges[-1]:
```

Comment:
It is randomly distributing the records within each range. If the range is too big, stream position can always be an indicator for grouping files. For example, initial file ordered also must be preserved as it was grouped by stream position earlier.

Reply:
Makes sense. I guess we will end up producing lot of small files as a result. The skipping untouched files also will ignore if the records aren't touched in a new delta. This could defeat the purpose of compaction altogether.

Reply:
I think that's why we need a compaction right after hot/cold repartitioning. Initial rebase compaction should normalize the file size for us

Reply:
n cold files and n hot files in case 1 seems incorrect. You are basically passing `max_records_per_entry` as None which will only create 1 file per range. So we may end up with only two files at the end. Can you confirm?

Reply:
you are right that one file per range, so in our case, given one date, e.g., ranges=[2023-01-01], each file will be split into two files, i.e., records <= 2023-01-01, and records>2023-01-01, so it'll be 2 files per original file